w3lib/url.py

```diff
@@ -95,10 +95,26 @@ def safe_url_string(
     decoded = to_unicode(url, encoding=encoding, errors="percentencode")
     parts = urlsplit(_ascii_tab_newline_re.sub("", decoded))
 
+    username, password, hostname, port_number = (
+        parts.username,
+        parts.password,
+        parts.hostname,
+        parts.port
+    )
+    netloc_bytes = b''
+
     # IDNA encoding can fail for too long labels (>63 characters)
     # or missing labels (e.g. http://.example.com)
     try:
-        netloc_bytes = parts.netloc.encode("idna")
+        # When we have hostname we use it instead of netloc directly
+        if hostname:
+            if isinstance(username, str) and isinstance(password, str):
+                netloc_bytes += f'{username}:{password}@'.encode("idna")
+            netloc_bytes += hostname.encode("idna")
+            if port_number:
+                netloc_bytes += f":{port_number}".encode("idna")
```

Comment:
You are encoding everything as `idna`, but the [StackOverflow](https://stackoverflow.com/questions/804336/best-way-to-convert-a-unicode-url-to-ascii-utf-8-percent-escaped-in-python/804380#804380) answer that @wRAR linked in the issue being addressed only does so for the `hostname`. Are you sure those other parts also need to be encoded as `idna`?

Reply:
Changed, thanks.